examples/pytorch/node2vec/model.py

```diff
@@ -0,0 +1,298 @@
+        neg_batch = batch.repeat(self.num_negatives)
+        neg_traces = torch.randint(self.N, (neg_batch.size(0), self.walk_length))
+        neg_traces = torch.cat([neg_batch.view(-1, 1), neg_traces], dim=-1)
+        neg_traces = neg_traces.unfold(1, self.window_size, 1)  # rolling window
+        neg_traces = neg_traces.contiguous().view(-1, self.window_size)
+
+        return pos_traces, neg_traces
+
+    def forward(self, nodes=None):
+        """
+        Returns the embeddings of the input nodes
+        Parameters
+        ----------
+        nodes: Tensor, optional
+            Input nodes, if set `None`, will return all the node embedding.
+
+        Returns
+        -------
+        Tensor
+            Node embedding
+
+        """
+        emb = self.embedding.weight
+        if nodes is None:
+            return emb
+        else:
+            return emb[nodes]
+
+    def loss(self, pos_trace, neg_trace):
+        """
+        Computes the loss given positive and negative random walks.
+        Parameters
+        ----------
+        pos_trace: Tensor
+            positive random walk trace
+        neg_trace: Tensor
+            negative random walk trace
+
+        """
+        e = 1e-15
+
+        # Positive
+        pos_start, pos_rest = pos_trace[:, 0], pos_trace[:, 1:].contiguous()  # start node and following trace
+        w_start = self.embedding(pos_start).unsqueeze(dim=1)
+        w_rest = self.embedding(pos_rest)
+        pos_out = (w_start * w_rest).sum(dim=-1).view(-1)
+
+        # Negative
+        neg_start, neg_rest = neg_trace[:, 0], neg_trace[:, 1:].contiguous()
+
+        w_start = self.embedding(neg_start).unsqueeze(dim=1)
+        w_rest = self.embedding(neg_rest)
+        neg_out = (w_start * w_rest).sum(dim=-1).view(-1)
+
+        # compute loss
+        pos_loss = -torch.log(torch.sigmoid(pos_out) + e).mean()
+        neg_loss = -torch.log(1 - torch.sigmoid(neg_out) + e).mean()
+
+        return pos_loss + neg_loss
+
+    def loader(self, batch_size):
+        """
+
+        Parameters
+        ----------
+        batch_size: int
+            batch size
+
+        Returns
+        -------
+        DataLoader
+            Node2vec training data loader
+
+        """
+        return DataLoader(torch.arange(self.N), batch_size=batch_size, shuffle=True, collate_fn=self.sample)
+
+    @torch.no_grad()
+    def evaluate(self, x_train, y_train, x_val, y_val):
+        """
+        Evaluate the quality of embedding vector via a downstream classification task with logistic regression.
+        """
+        x_train = self.forward(x_train)
+        x_val = self.forward(x_val)
+
+        x_train, y_train = x_train.cpu().numpy(), y_train.cpu().numpy()
+        x_val, y_val = x_val.cpu().numpy(), y_val.cpu().numpy()
+        lr = LogisticRegression(solver='lbfgs', multi_class='auto', max_iter=150).fit(x_train, y_train)
+
+        return lr.score(x_val, y_val)
+
+
+class Node2vecModel(object):
+    """
```

Comment:
I ultimately decided to keep it this way because later on people can just change the Node2vec module (e.g. replacing embedding with MLP, etc.) without changing the trainer.